internal/io/sql/source.go

```diff
@@ -0,0 +1,208 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pingcap/failpoint"
+
+	"github.com/lf-edge/ekuiper/contract/v2/api"
+	"github.com/lf-edge/ekuiper/v2/internal/io/sql/client"
+	"github.com/lf-edge/ekuiper/v2/pkg/cast"
+	"github.com/lf-edge/ekuiper/v2/pkg/connection"
+	util "github.com/lf-edge/ekuiper/v2/pkg/dbpool"
+	"github.com/lf-edge/ekuiper/v2/pkg/dbpool/sqldatabase/sqlgen"
+	"github.com/lf-edge/ekuiper/v2/pkg/timex"
+)
+
+type SQLSourceConnector struct {
+	conf  *SQLConf
+	Query sqlgen.SqlQueryGenerator
+	conn  *client.SQLConnection
+	props map[string]any
+}
+
+type SQLConf struct {
+	Interval cast.DurationConf `json:"interval"`
+	DBUrl    string            `json:"dburl"`
+}
+
+func (s *SQLSourceConnector) Provision(ctx api.StreamContext, props map[string]any) error {
+	cfg := &SQLConf{}
+	err := cast.MapToStruct(props, cfg)
+	failpoint.Inject("MapToStructErr", func() {
+		err = errors.New("MapToStruct")
+	})
+	if err != nil {
+		return fmt.Errorf("read properties %v fail with error: %v", props, err)
+	}
+	if time.Duration(cfg.Interval) < 1 {
+		return fmt.Errorf("interval should be defined")
+	}
+	if len(cfg.DBUrl) < 1 {
+		return fmt.Errorf("dburl should be defined")
+	}
+	s.conf = cfg
+	s.props = props
+	sqlDriver, err := util.ParseDriver(cfg.DBUrl)
+	if err != nil {
+		return fmt.Errorf("dburl.Parse %s fail with error: %v", cfg.DBUrl, err)
+	}
+	generator, err := sqlgen.GetQueryGenerator(sqlDriver, props)
+	failpoint.Inject("GetQueryGeneratorErr", func() {
+		err = errors.New("GetQueryGeneratorErr")
+	})
+	if err != nil {
+		return fmt.Errorf("GetQueryGenerator %s fail with error: %v", cfg.DBUrl, err)
+	}
+	s.Query = generator
+	return nil
+}
+
+func (s *SQLSourceConnector) Connect(ctx api.StreamContext) error {
+	ctx.GetLogger().Infof("Connecting to sql server")
+	var cli *client.SQLConnection
+	var err error
+	id := fmt.Sprintf("%s-%s-%s-sql-source", ctx.GetRuleId(), ctx.GetOpId(), s.conf.DBUrl)
+	conn, err := connection.FetchConnection(ctx, id, "sql", s.props)
+	if err != nil {
+		return err
+	}
+	cli = conn.(*client.SQLConnection)
+	s.conn = cli
+	return err
+}
+
+func (s *SQLSourceConnector) Close(ctx api.StreamContext) error {
+	ctx.GetLogger().Infof("Closing sql source connector url:%v", s.conf.DBUrl)
+	if s.conn != nil {
+		id := fmt.Sprintf("%s-%s-%s-sql-source", ctx.GetRuleId(), ctx.GetOpId(), s.conf.DBUrl)
+		connection.DetachConnection(ctx, id, s.props)
+		s.conn.DetachSub(ctx, s.props)
+	}
+	return nil
+}
+
+func (s *SQLSourceConnector) Subscribe(ctx api.StreamContext, ingest api.TupleIngest, ingestError api.ErrorIngest) error {
```

Comment:
Try to implement PullTupleSource. The source  can only focus on how to pull, do not need to take care about the ticker.